tests/functional_tests/shmem_ptr_tester.cpp

```diff
@@ -31,59 +31,146 @@ using namespace rocshmem;
 /******************************************************************************
  * DEVICE TEST KERNEL
  *****************************************************************************/
-__global__ void ShmemPtrTest(char *r_buf, int *available) {
+__global__ void ShmemPtrTest(int loop, int skip, long long int *start_time,
+                             long long int *end_time, char *dest, int wf_size,
+                             ShmemContextType ctx_type, int *available) {
+  __shared__ rocshmem_ctx_t ctx;
+  int wg_id = get_flat_grid_id();
+  int t_id  = get_flat_block_id();
+  int wf_id = t_id / wf_size;
+
   rocshmem_wg_init();
+  rocshmem_wg_ctx_create(ctx_type, &ctx);
+
+  /**
+   * Shared array to capture the start time for each wavefront
+   * Max threads per block = 1024, wavefront size = 64 (in most GPUs)
+   * Maximum array size required = 1024/64 = 16
+   */
+  __shared__ long long int wf_start_time[16];
+
+
+  /**
+   * Calculate start index for each thread within the grid
+   */
+  dest += get_flat_id();
+
+  char *local_addr = dest;
+  void *remote_addr = rocshmem_ptr((void *)local_addr, 1);
+  if (remote_addr != NULL) {
+    *available = 1;
+  }
+
+  if(*available) {
+    for (int i = 0; i < loop + skip; i++) {
+      if (i == skip) {
+        __syncthreads();
+        // Ensures all RMA calls from the skip loops are completed
+        if(is_thread_zero_in_block()) {
+          rocshmem_ctx_quiet(ctx);
+        }
+        __syncthreads();
+        // Capture the start time of each wavefront to identify the earliest one
+        wf_start_time[wf_id] = wall_clock64();
+      }
 
-  if (hipThreadIdx_x == 0) {
-    char *local_addr = r_buf + 4;
-    void *remote_addr = rocshmem_ptr((void *)local_addr, 1);
-    if (remote_addr != NULL) {
-      *available = 1;
       ((char *)remote_addr)[0] = '1';
     }
   }
 
+  __syncthreads();
+  if(is_thread_zero_in_block()) {
+    rocshmem_ctx_quiet(ctx);
+  }
+
+  /**
+   * End time of the last wavefront is recorded by overwriting
+   * the value previously set by earlier wavefronts.
+   */
+  end_time[wg_id] = wall_clock64();
+
+  // Find the earliest start time
+  int num_wfs = (get_flat_block_size() - 1 ) / wf_size + 1;
+  for (int i = num_wfs / 2; i > 0; i >>= 1 ) {
+    if(t_id < i) {
+      wf_start_time[t_id] = min(wf_start_time[t_id], wf_start_time[t_id + i]);
+    }
+  }
+
+  // For data validation in remote PE
+  if(*available == 1 && get_flat_id() == 0) {
+    local_addr = dest + get_flat_grid_size();
+    int *_available = (int*)rocshmem_ptr((void *)local_addr, 1);
```

Comment:
```suggestion
    shmem_put(local_addr, 1, 1)
```
this will prevent the test from failing silent if the visibility updates from rocshmem_ptr are not enforced